mirgecom/inviscid.py

```diff
@@ -82,21 +82,25 @@ def get_inviscid_timestep(discr, eos, cfl, q):
     Currently, it's a hack waiting for the geometric_factor helpers port
     from grudge.
     """
+    from mpi4py import MPI
     dim = discr.dim
-    mesh = discr.mesh
     order = max([grp.order for grp in discr.discr_from_dd("vol").groups])
-    nelements = mesh.nelements
-    nel_1d = nelements ** (1.0 / (1.0 * dim))
-
-    # This roughly reproduces the timestep AK used in wave toy
-    dt = (1.0 - 0.25 * (dim - 1)) / (nel_1d * order ** 2)
-    return cfl * dt
-
-#    dt_ngf = dt_non_geometric_factor(discr.mesh)
-#    dt_gf  = dt_geometric_factor(discr.mesh)
-#    wavespeeds = compute_wavespeed(w,eos=eos)
-#    max_v = clmath.max(wavespeeds)
-#    return c*dt_ngf*dt_gf/max_v
+    cv = split_conserved(dim, q)
+
+    import grudge.op as op
+    h_min_local = op.h_min_from_volume(discr) / (order * order)
+    from mirgecom.fluid import compute_wavespeed
+    local_wavespeeds = compute_wavespeed(dim, eos, cv)
+    max_wavespeed_local = op.nodal_max(discr, "vol", local_wavespeeds)
+
+    mpi_comm = discr.mpi_communicator
+    if mpi_comm is None:
+        return cfl * h_min_local / max_wavespeed_local
+
+    h_min_global = mpi_comm.allreduce(h_min_local, op=MPI.MIN)
+    max_wavespeed_global = mpi_comm.allreduce(max_wavespeed_local, op=MPI.MAX)
+
+    return cfl * h_min_global / max_wavespeed_global
```

Comment:
I think this is kind of dangerous. It's perhaps better than nothing, but it'll fail for elements with large aspect ratio. @MTCam mentioned a more robust procedure given in the JSH/TW book, which I've had good experience with. Question: Should we merge this as a stopgap until we have that, or go for the actual article.

Reply:
@inducer Where is the "more robust procedure"? @MTCam and I spend some time trying to find it. This approach here is basically the formula given on page 325, equation 7.32

Reply:
I def remember reading it, but @thomasgibson and I searched together and could not dig it up.

Reply:
The 1D approach is towards the end of Section 4.7 for me:

![grafik](https://user-images.githubusercontent.com/352067/120031239-29bee580-bfbe-11eb-9a29-9e800f658861.png)

The generalization to more dimensions is in Section 6.4:

![grafik](https://user-images.githubusercontent.com/352067/120031407-6094fb80-bfbe-11eb-9d34-dc8576b4e6dc.png)

Also
- https://github.com/tcew/nodal-dg/blob/master/Codes1.1/Codes2D/dtscale2D.m
- https://github.com/tcew/nodal-dg/blob/master/Codes1.1/Codes3D/dtscale3D.m

Reply:
Personally, I like the expression Hesthaven uses here:

<img width="403" alt="Screen Shot 2021-05-28 at 3 59 39 PM" src="https://user-images.githubusercontent.com/8116930/120041372-b91fc500-bfcd-11eb-8c5a-57746dcc30bf.png">

It has the advantage that it comes out in units of time, and is easy to understand.  It doesn't need to be very precise - it's just a guide and the user will need to tune dt or cfl to the particular simulation anyway.

Reply:
They're not incompatible though. The approach above just provides more precise C's to the formula you mention.

Reply:
I will implement whatever you guys think is best.  

My main complaints about the other formula is that it is hard to understand, and is not clear to me how to derive it for 3D (or even 2D how it's presented), and it does not have the correct units.  Since it needs to be tuned at runtime no matter what, in my opinion understandability and proper units wins.

If grudge can provide the min del R and min R_d, we can whip it out.

Reply:
But also, ugh, the 7.32 formula at least has some consideration of the viscous terms, and seems specific to viscous flow.  I like that.

Reply:
> I think this is kind of dangerous. It's perhaps better than nothing, but it'll fail for elements with large aspect ratio.

@inducer I can see the approach you linked perhaps gives a tighter bound, but how is this any less dangerous? The definition of `r_D` uses the cell area / perimeter. This is just as dangerous for degenerate elements.

> it'll fail for elements with large aspect ratio

Can you clarify here what you mean? In the atmosphere, we deal with "large aspect ratios" all the time. What is "large" here? ~O(100)? ~O(1000)?

Reply:
Maybe we should be using the circumradius in our calculations? https://mathworld.wolfram.com/Circumsphere.html

Reply:
All I'm saying is that I'd like something that vaguely responds to the element aspect ratio, with a test that it "catches" the growth of the eigenvalues of DG differentiation under modest amounts of h and p refinement. Aspect ratio is easy to test by "squishing' (affine-mapping) a rect mesh.

Reply:
This comment is rather concerning (for the 3D case): https://github.com/tcew/nodal-dg/blob/master/Codes1.1/Codes3D/Maxwell3D.m#L14

Reply:
Found this paper: https://arxiv.org/pdf/1806.03095.pdf, where they choose a timestep that is more or less what @MTCam posted (equation 7.32 in H&W). 
![image](https://user-images.githubusercontent.com/14180421/120086618-f0fa3b80-c0a5-11eb-9340-3918687b6d0d.png)

Not necessarily claiming it's better, just that it's at least taking into account the viscous part. One thing I'm concerned about with what @inducer posted is how the viscous timestep should be handled. I've started coding it up, but I'm still not totally convinced it's better.

Reply:
Should `2/3 min delta r` replace the `1/N**2` term in the bound and `h_min` is the circumscribed radius?

Reply:
Uhg DG + time-stepping is so damn ugly

Reply:
> Should `2/3 min delta r` replace the `1/N**2` term in the bound and `h` is the circumscribed radius?

I'd say so. (Are you pursuing the thing with the singular values of the mapping Jacobian we had discussed any further, as opposed to the incircle or circumcircle radius?)

Reply:
> Uhg DG + time-stepping is so damn ugly

:shrug: You're guessing eigenvalues. At a cost, you could also use adaptive time stepping and forget about CFL estimation.

Reply:
> (Are you pursuing the thing with the singular values of the mapping Jacobian we had discussed any further, as opposed to the incircle or circumcircle radius?)

I going to try that, but I'm struggling to get the right matrix per element. I think I should be using https://github.com/inducer/grudge/blob/main/grudge/geometry/metrics.py#L180, but I need to reshape the components and get it into the form of a matrix. My plan from there was to do a SVD calculation and take the minimum singular value.

Reply:
https://github.com/inducer/grudge/pull/107

Reply:
(I.e.: prototyped something)

Reply:
Took a stab at extending @inducer's prototype and cleaned up the functions: https://github.com/inducer/grudge/pull/108

Reply:
OK, I'm happy now with https://github.com/inducer/grudge/pull/108: @MTCam, this should make the `dt_non_geometric_factor`/`dt_geometric_factor` usable for this PR. Don't worry about the viscous case; these two factors are designed to give you an approximation to `h_min / N^2`, which means the only difference in the expression you posted that handles the viscous timestep is that you replace `h / N^2` with the product of the nongeometric and geometric factors.

Reply:
Wow, that's cool!  Thanks a ton @thomasgibson and @inducer!